src/execution_tree/primitives/sub_operation.cpp

```diff
@@ -0,0 +1,206 @@
+
+        if (lhs_size != rhs_size)
+        {
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "sub_operation::sub2d2d",
+                "the dimensions of the operands do not match");
+        }
+
+        using array_type = Eigen::Array<double, Eigen::Dynamic, Eigen::Dynamic>;
+        using result_type = Eigen::Matrix<double, Eigen::Dynamic, Eigen::Dynamic>;
+        using array_map_type = Eigen::Map<array_type const>;
+
+        auto lambda = [&]()->result_type {
+            auto val1 = *ops.begin();
+            auto val2 = *(++ops.begin());
+            array_map_type first(val1.data(), lhs_size[0], lhs_size[1]);
+            array_map_type second(val2.data(), lhs_size[0], lhs_size[1]);
+            return first - second;
+        };
+        result_type result = lambda();
+        return ir::node_data<double>(std::move(result));
+    }
+
+    ir::node_data<double> sub_operation::sub2d(operands_type const& ops) const
+    {
+        std::size_t rhs_dims = ops[1].num_dimensions();
+        switch(rhs_dims)
+        {
+        case 2:
+            return sub2d2d(ops);
+
+        case 0: HPX_FALLTHROUGH;
+        case 1: HPX_FALLTHROUGH;
+        default:
+            HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                "sub_operation::sub2d",
+                "the operands have incompatible number of dimensions");
+        }
+    }
+
+    namespace detail
+    {
+        template <typename T, typename F>
+        auto map(std::vector<T> const& in, F && f)
+        ->  std::vector<decltype(hpx::util::invoke(f, std::declval<T>()))>
+        {
+            std::vector<decltype(hpx::util::invoke(f, std::declval<T>()))> out;
+            out.reserve(in.size());
+            for (auto const& v : in)
+            {
+                out.push_back(hpx::util::invoke(f, v));
+            }
+            return out;
+        }
+    }
+
+    // implement '-' for all possible combinations of lhs and rhs
+    hpx::future<ir::node_data<double>> sub_operation::eval() const
+    {
+        return hpx::dataflow(hpx::util::unwrapping(
+            [this](std::vector<ir::node_data<double>> && ops)
+            {
+                std::size_t lhs_dims = ops[0].num_dimensions();
+                switch (lhs_dims)
+                {
+                case 0:
+                    return sub0d(ops);
+
+                case 1:
+                    return sub1d(ops);
+
+                case 2:
+                    return sub2d(ops);
+
+                default:
+                    HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                        "sub_operation::eval",
+                        "left hand side operand has unsupported number of "
+                        "dimensions");
+                }
+            }),
+            detail::map(operands_,
+                [](primitive const& p) -> hpx::future<ir::node_data<double>>
+                {
+                    return p.eval();
+                })
+        );
+    }
+}}}
```

Comment:
This implementation will support working with exectly 2 operands only. Do you think you could extend it to work with an arbitrary amount of operands (>= 2)?

Reply:
Yes, I just pushed a version that should be able support more than 2 versions in e8f78115de03af2ca03566ebeb1ca4e454285503.